File: app/src/main/java/com/example/fellow_traveller/Chat/MessagesAdapter.java
package com.example.fellow_traveller.Chat;

import android.content.Context;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.example.fellow_traveller.MainActivity;
import com.example.fellow_traveller.Models.GlobalClass;
import com.example.fellow_traveller.R;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.squareup.picasso.Picasso;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import de.hdodenhof.circleimageview.CircleImageView;

public class MessagesAdapter extends RecyclerView.Adapter<MessagesAdapter.MessagesViewHolder> {
    private ArrayList<MessageItem> messagesList;
    private Context myContext;
    public static final int SEND_SINGLE = 0;
    public static final int SEND_TOP = 1;
    public static final int SEND_MIDDLE = 2;
    public static final int SEND_BOTTOM = 3;
    public static final int RECEIVE_SINGLE = 4;
    public static final int RECEIVE_TOP = 5;
    public static final int RECEIVE_MIDDLE = 6;
    public static final int RECEIVE_BOTTOM = 7;
    private GlobalClass globalClass;
    private int myId;


    public static class MessagesViewHolder extends RecyclerView.ViewHolder {
        public TextView text;
        public TextView name;
        public TextView date;
        public CircleImageView userImage;


        public MessagesViewHolder(@NonNull View itemView) {
            super(itemView);

            text = itemView.findViewById(R.id.message);
            name = itemView.findViewById(R.id.sender_message);
            date = itemView.findViewById(R.id.message_time);
            userImage = itemView.findViewById(R.id.image_message);
        }
    }

    public MessagesAdapter(ArrayList<MessageItem> messList, Context context) {
        messagesList = messList;
        myContext = context;

    }

    @NonNull
    @Override
    public MessagesViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == SEND_SINGLE) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.message_send_single, parent, false);
            MessagesViewHolder evh = new MessagesViewHolder(v);
            return evh;
        } else if (viewType == SEND_BOTTOM) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.message_send_bottom, parent, false);
            MessagesViewHolder evh = new MessagesViewHolder(v);
            return evh;
        } else if (viewType == SEND_MIDDLE) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.message_send_middle, parent, false);
            MessagesViewHolder evh = new MessagesViewHolder(v);
            return evh;
        } else if (viewType == SEND_TOP) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.message_send_top, parent, false);
            MessagesViewHolder evh = new MessagesViewHolder(v);
            return evh;
        }else if (viewType == RECEIVE_TOP) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.message_receive_top, parent, false);
            MessagesViewHolder evh = new MessagesViewHolder(v);
            return evh;
        } else if (viewType == RECEIVE_BOTTOM) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.message_receive_bottom, parent, false);
            MessagesViewHolder evh = new MessagesViewHolder(v);
            return evh;
        }else if (viewType == RECEIVE_MIDDLE) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.message_receive_middle, parent, false);
            MessagesViewHolder evh = new MessagesViewHolder(v);
            return evh;
        }else {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.message_receive_single, parent, false);
            MessagesViewHolder evh = new MessagesViewHolder(v);
            return evh;
        }
    }

    @Override
    public void onBindViewHolder(@NonNull final MessagesViewHolder  holder, int position) {
       final MessageItem currentItem = messagesList.get(position);

        holder.text.setText(currentItem.getText());

        Date currentDate = new Date(currentItem.getTimestamp()*1000);
        DateFormat dateFormat = convertDateFormat(currentItem.getTimestamp()*1000);
        holder.date.setText(dateFormat.format(currentDate));

        final int viewType = getItemViewType(position);


        //Retrieve the id of sender of the message, to get the sender's username
        DatabaseReference userName = FirebaseDatabase.getInstance().getReference("Users").child(Integer.toString(currentItem.getId()));

        userName.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                String nameFirebase = (String) dataSnapshot.child("name").getValue(String.class);
                String userImage = (String) dataSnapshot.child("image").getValue(String.class);
                holder.name.setText(nameFirebase);

                //Get the image URL and check the viewType, if message is type received then try to load senders image
                if(viewType == 4 || viewType == 5 || viewType == 6 || viewType == 7){
                    try {
                        Picasso.get().load(userImage).into(holder.userImage);
                    }
                    catch(Exception e) {
                        //  Block of code to handle errors
                    }
                }

            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });




    }

    private DateFormat convertDateFormat(long myTimestamp) {

        //Get The year and date of our timestamp and from current year
        Calendar cal = Calendar.getInstance();
        cal.setTimeInMillis(myTimestamp);
        int compareYear = cal.get(Calendar.YEAR);
        int compareDay = cal.get(Calendar.DATE);
        int thisYear = Calendar.getInstance().get(Calendar.YEAR);
        int thisDate = Calendar.getInstance().get(Calendar.DATE);

        DateFormat dateFormat;

        if(thisDate==compareDay){
            dateFormat = new SimpleDateFormat("h:mm a");
            return dateFormat;
        }else if(thisYear==compareYear){
            dateFormat = new SimpleDateFormat("EEE, d MMM - h:mm a");
            return dateFormat;
        }else{
            dateFormat = new SimpleDateFormat("d MMM yyyy - h:mm a");
            return dateFormat;
        }
    }

    @Override
    public int getItemCount() {
        return messagesList.size();
    }

    @Override
    public int getItemViewType(int position) {
        MessageItem curr = messagesList.get(position);
        int current = curr.getId();
        int currNext = -1;
        int currPrev = -1;

        //Get myId from Global Class
        globalClass = (GlobalClass) myContext.getApplicationContext();
        myId = globalClass.getCurrentUser().getId();
        //Initialize and check if we run out of the list
        if (position + 1 < messagesList.size()) {
            currNext = messagesList.get(position + 1).getId();
        }
        if (position - 1 > -1) {
            currPrev = messagesList.get(position - 1).getId();
        }
        //Finished initialization


        if (current == myId) {

            if ((currNext == current) && (current == currPrev)) {
                return SEND_MIDDLE;
            } else if ((current != currNext) && (current == currPrev)) {
                return SEND_BOTTOM;
            } else if ((current != currPrev) && (current == currNext)) {
                return SEND_TOP;
            } else { return SEND_SINGLE;
                }
        }else {

            if ((currNext == current) && (current == currPrev)) {
                return RECEIVE_MIDDLE;
            } else if ((current != currNext) && (current == currPrev)) {
                return RECEIVE_BOTTOM;
            } else if ((current != currPrev) && (current == currNext)) {
                return RECEIVE_TOP;
            } else {
                return RECEIVE_SINGLE;
            }
        }


    }

}

File: app/src/main/java/gr/fellow/fellow_traveller/ui/home/adapter/MessagesAdapter.kt
package gr.fellow.fellow_traveller.ui.home.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import gr.fellow.fellow_traveller.R
import gr.fellow.fellow_traveller.domain.chat.ChatMessage
import gr.fellow.fellow_traveller.domain.user.UserInfo
import gr.fellow.fellow_traveller.ui.extensions.loadImageFromUrl
import gr.fellow.fellow_traveller.ui.extensions.setTextHtml
import gr.fellow.fellow_traveller.utils.convertTimestampToFormat

class MessagesAdapter(
    private val messagesList: MutableList<ChatMessage>,
    private val myId: String,
    private val participantsInfo: MutableList<UserInfo> = mutableListOf<UserInfo>(),

    ) : RecyclerView.Adapter<MessagesAdapter.ViewHolder>() {

    private val MESSAGE_TOP = 0
    private val MESSAGE_MIDDLE = 1
    private val MESSAGE_ENTRY = 2
    private val MESSAGE_EXITED = 3
    private val MESSAGE_FIRST = 4

    private companion object {
        private const val TIME_DIFFERENCE: Long = 300000
    }

    // private var viewTypeLayout = -1


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        if (viewType == MESSAGE_TOP) {
            // viewTypeLayout = 0
            val view = LayoutInflater.from(parent.context).inflate(R.layout.message_item_top, parent, false)
            return ViewHolder(view, 0)
        } else if (viewType == MESSAGE_MIDDLE) {
            //viewTypeLayout = 1
            val view = LayoutInflater.from(parent.context).inflate(R.layout.message_item_middle, parent, false)
            return ViewHolder(view, 1)
        } else if (viewType == MESSAGE_ENTRY) {
            //viewTypeLayout = 2
            val view = LayoutInflater.from(parent.context).inflate(R.layout.message_item_entry, parent, false)
            return ViewHolder(view, 2)
        } else if (viewType == MESSAGE_EXITED) {
            //viewTypeLayout = 3
            val view = LayoutInflater.from(parent.context).inflate(R.layout.message_item_exited, parent, false)
            return ViewHolder(view, 3)
        } else {
            //viewTypeLayout = 4
            val view = LayoutInflater.from(parent.context).inflate(R.layout.message_item_first, parent, false)
            return ViewHolder(view, 4)
        }
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {

        val currentItem = messagesList[position]

        /** Get the background shape for each message **/
        if (currentItem.messageType == 0) {
            if (currentItem.senderId == myId)
                holder.message.setBackgroundResource(R.drawable.message_shape_sended)
            else
                holder.message.setBackgroundResource(R.drawable.message_shape_received)

        } else if (currentItem.messageType == 1) {
            holder.message.setBackgroundResource(R.drawable.message_shape_entry)
        } else
            holder.message.setBackgroundResource(R.drawable.message_shape_exited)


        /** Load details for each message **/
        //if viewType is MESSAGE_TOP load all the details
        when {
            currentItem.messageType == 0 -> {
                holder.message.text = currentItem.text
                try {
                    val user = participantsInfo.first {
                        it.id == currentItem.senderId
                    }
                    holder.name.text = user.firstName
                    holder.image.loadImageFromUrl(user.picture)
                } catch (e: NoSuchElementException) {
                    holder.name.text = currentItem.senderName
                    holder.image.loadImageFromUrl(currentItem.senderImage)
                }

                holder.date.text = convertTimestampToFormat(currentItem.timestamp, "hh:mm aaa")
            }
            //Entry Layout
            holder.viewTypeLayout == 2 -> holder.message.setTextHtml(holder.message.context.getString(R.string.conversation_passenger_entry, currentItem.senderName))

            //Exited Layout
            holder.viewTypeLayout == 3 -> holder.message.setTextHtml(holder.message.context.getString(R.string.conversation_passenger_exited, currentItem.senderName))

            //First Message Layout
            holder.viewTypeLayout == 4 -> holder.image.loadImageFromUrl(currentItem.senderImage)

            else -> {
                holder.message.text = currentItem.text
            }
        }

    }


    class ViewHolder(view: View, aViewTypeLayout: Int) : RecyclerView.ViewHolder(view) {
        val message: TextView = view.findViewById(R.id.message)
        val name: TextView = view.findViewById(R.id.sender_name)
        val date: TextView = view.findViewById(R.id.message_time)
        val image: ImageView = view.findViewById(R.id.sender_image)

        val viewTypeLayout = aViewTypeLayout
    }

    override fun getItemViewType(position: Int): Int {

        var currNext = "-1"
        var currPrev = "-1"
        val current = messagesList[position].senderId
        //Initialize and check if we run out of the list
        if (position + 1 < messagesList.size) {
            currNext = messagesList[position + 1].senderId;
        } else {
            currNext = "-1"
        }

        if (position - 1 > -1) {
            currPrev = messagesList[position - 1].senderId;
        } else {
            currPrev = "-1"
        }

        /** FROM FIREBASE WE GET **/

        /** 0 for MESSAGE_TOP and MESSAGE_MIDDLE **/
        /** 1 for MESSAGE_ENTRY **/
        /** 2 for MESSAGE_EXITED **/
        /** 3 for MESSAGE_FIRST **/


        if (messagesList[position].messageType == 0) {
            if ((current == currPrev)) {
                if (messagesList[position - 1].messageType == 1 || messagesList[position - 1].messageType == 3) {
                    return MESSAGE_TOP
                } else {
                    if (messagesList[position].timestamp - messagesList[position - 1].timestamp > TIME_DIFFERENCE)
                        return MESSAGE_TOP
                    else
                        return MESSAGE_MIDDLE
                }
            } else {
                return MESSAGE_TOP;
            }
        } else if (messagesList[position].messageType == 1) {
            return MESSAGE_ENTRY
        } else if (messagesList[position].messageType == 2) {
            return MESSAGE_EXITED
        } else
            return MESSAGE_FIRST

    }


    override fun getItemCount() = messagesList.size

}